WalletWasabi.Gui/MainWindow.xaml.cs

```diff
@@ -78,74 +87,114 @@ private async void MainWindow_ClosingAsync(object? sender, CancelEventArgs e)
 			}
 		}
 
-		private async Task ClosingAsync()
+		/// <summary>
+		/// This will try to close the main window of the application.
+		/// </summary>
+		/// <param name="tryToDequeue">If true then the user will get a dialog where a status of the dequing process is shown and the window will be kept open until it successfully finishes. If false then the dequeing process will be handled later at the disposal of the business logic.</param>
+		/// <returns></returns>
+		public async Task ClosingAsync(bool tryToDequeue)
 		{
-			bool closeApplication = false;
+			if (IsClosed)
+			{
+				return;
+			}
+
+			if (Dispatcher.UIThread?.CheckAccess() is false)
+			{
+				// We are not on the UI thread, let's synchronize.
+				await Dispatcher.UIThread.InvokeAsync(async () =>
+				{
+					// Safety check if it was closed meanwhile synchronization.
+					if (IsClosed)
+					{
+						return;
+					}
+
+					await ClosingAsync(tryToDequeue).ConfigureAwait(false);
+				}).ConfigureAwait(false);
+				return;
+			}
+
+			// We are on the UI Thread now, do not use ConfigureAwait(false) after this line.
+
+			bool closeApplication = true;
 			try
 			{
 				// Indicate -> do not add any more alices to the coinjoin.
 				Global.WalletManager.SignalQuitPending(true);
-
-				if (Global.WalletManager.AnyCoinJoinInProgress())
+				if (tryToDequeue && Global.WalletManager.AnyCoinJoinInProgress())
 				{
 					var dialog = new CannotCloseDialogViewModel();
 
-					closeApplication = await MainWindowViewModel.Instance.ShowDialogAsync(dialog); // start the deque process with a dialog
+					// Start the deque process with a dialog.
+					closeApplication = await MainWindowViewModel.Instance.ShowDialogAsync(dialog);
 				}
-				else
+
+				if (!closeApplication)
 				{
-					closeApplication = true;
+					//The user aborted the close, let's go to finally.
+					return;
 				}
 
-				if (closeApplication)
+				try
 				{
-					try
-					{
-						if (Global.UiConfig is { }) // UiConfig not yet loaded.
-						{
-							Global.UiConfig.WindowState = WindowState;
+					SaveUiConfig();
 
-							IDocumentTabViewModel? selectedDocument = IoC.Get<IShell>().SelectedDocument;
-							Global.UiConfig.LastActiveTab = selectedDocument is null
-								? nameof(HistoryTabViewModel)
-								: selectedDocument.GetType().Name;
+					Hide();
 
-							Global.UiConfig.ToFile();
-							Logger.LogInfo($"{nameof(Global.UiConfig)} is saved.");
-						}
-
-						Hide();
-						var wm = IoC.Get<IShell>().Documents?.OfType<WalletManagerViewModel>().FirstOrDefault();
-						if (wm is { })
+					if (IoC.Get<IShell>().Documents is { } docs)
+					{
+						foreach (var doc in docs.OfType<WasabiDocumentTabViewModel>().ToArray())
```

Comment:
```suggestion
						foreach (var doc in docs.OfType<WasabiDocumentTabViewModel>())
```

Looks redundant to me.

Reply:
ACK